eta/core/primitives.py

```diff
@@ -0,0 +1,605 @@
+                flow_polar = np.dstack((mag, ang))
+
+                if polar_path:
+                    # Write polar fields
+                    np.save(polar_path % p.frame_number, flow_polar)
+
+                if vid_path:
+                    # Write flow visualization frame
+                    p.write(_polar_flow_to_img(mag, ang))
+
+    def _process_frame(self, img):
+        '''Processes the next frame.
+
+        Args:
+            img: the next frame
+
+        Returns:
+            flow: the optical flow for the frame in Cartesian coordinates
+        '''
+        raise NotImplementedError("subclass must implement _process_frame()")
+
+    def _reset(self):
+        '''Prepares the object to start processing a new video.'''
+        pass
+
+
+def _polar_flow_to_img(mag, ang):
+    hsv = np.zeros(mag.shape + (3,), dtype=mag.dtype)
+    hsv[..., 0] = (89.5 / np.pi) * ang  # [0, 179]
+    hsv[..., 1] = 255
+    #hsv[..., 2] = np.minimum(255 * mag, 255)  # [0, 255]
+    hsv[..., 2] = cv2.normalize(mag, None, 0, 255, cv2.NORM_MINMAX)  # [0, 255]
+    return cv2.cvtColor(hsv.astype(np.uint8), cv2.COLOR_HSV2BGR)
+
+
+class FarnebackDenseOpticalFlow(DenseOpticalFlow):
+    '''Computes dense optical flow on a video using Farneback's method.
+
+    This class is a wrapper around the OpenCV `calcOpticalFlowFarneback`
+    function.
+    '''
+
+    def __init__(
+            self,
+            pyramid_scale=0.5,
+            pyramid_levels=3,
+            window_size=15,
+            iterations=3,
+            poly_n=7,
+            poly_sigma=1.5,
+            use_gaussian_filter=False):
+        '''Constructs a FarnebackDenseOpticalFlow object.
+
+        Args:
+            pyramid_scale (0.5): the image scale (<1) to build pyramids for
+                each image
+            pyramid_levels (3): number of pyramid layers including the initial
+                image
+            window_size (15): averaging window size
+            iterations (3): number of iterations to perform at each pyramid
+                level
+            poly_n (7): size of the pixel neighborhood used to find polynomial
+                expansion in each pixel
+            poly_sigma (1.5): standard deviation of the Gaussian that is used
+                to smooth derivatives
+            use_gaussian_filter (False): whether to use a Gaussian filter
+                instead of a box filer
+        '''
+        self.pyramid_scale = pyramid_scale
+        self.pyramid_levels = pyramid_levels
+        self.window_size = window_size
+        self.iterations = iterations
+        self.poly_n = poly_n
+        self.poly_sigma = poly_sigma
+        self.use_gaussian_filter = use_gaussian_filter
+
+        self._prev_frame = None
+        self._flags = (
+            cv2.OPTFLOW_FARNEBACK_GAUSSIAN if use_gaussian_filter else 0)
+
+    def _process_frame(self, img):
+        curr_frame = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
```

Comment:
ACK.  We recently decided to establish and enforce a contract that all images ETA reads will be converted to RGB from BGR (when ingested as such).  This call violates that contract.  (Of course, this is color-to-gray, so it is not fatal, but nevertheless should be fixed.)

Reply:
I will do this in a separate PR. As currently written, ETA loads images in BGR format